### PR TITLE
Fix serializing complex numbers with negative zero imaginary part

### PR DESCRIPTION
**Pull request by lhnwrk**
## Change Summary

Complex serializer currently dumps negative zero and NaN imaginary part into the wrong format:
```py
from pydantic import TypeAdapter

adapter = TypeAdapter(complex)
assert adapter.dump_json(complex(1, -0.0)) == b'"1-0j"'  # got b'"1+-0j"' instead
assert adapter.dump_json(complex(1, float("nan"))) == b'"1+NaNj"'  # got b'"1NaNj"' instead
```

This PR:
- Handle negative zero or NaN imaginary part correctly.
- Add tests for negative zero or NaN imaginary part.

## Related issue number

N/A

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by lhnwrk** [comment]
PR is ready and all checks have passed, please review 🙌

File: pydantic-core/src/serializers/type_serializers/complex.rs [hunk 1]
```diff
@@ -80,7 +80,7 @@ pub fn complex_to_str(py_complex: &Bound<'_, PyComplex>) -> String {
     let mut s = format!("{im}j");
     if re != 0.0 {
         let mut sign = "";
-        if im >= 0.0 {
+        if im.is_nan() || im.is_sign_positive() {
```
Comment (davidhewitt) [inline pydantic-core/src/serializers/type_serializers/complex.rs:86]: Is the `is_nan()` component necessary? It looks like the [docs for `is_sign_positive`](https://docs.rs/num-traits/0.2.19/num_traits/float/trait.Float.html#tymethod.is_sign_positive) suggest that positive NaN is included in that check.

File: pydantic-core/tests/serializers/test_complex.py [hunk 2]
```diff
@@ -36,7 +39,7 @@ def test_complex_json(value, expected):
     if math.isnan(value.real):
         assert math.isnan(c.real)
     else:
-        assert c.imag == value.imag
+        assert c.real == value.real
```
Comment (davidhewitt) [inline pydantic-core/tests/serializers/test_complex.py:42]: 👍

**Review by davidhewitt — COMMENTED** [review]
Thanks for the PR! One brief question, otherwise looking good to me.

File: pydantic-core/src/serializers/type_serializers/complex.rs (hunk 1 above)
Comment (lhnwrk) [reply pydantic-core/src/serializers/type_serializers/complex.rs:86]: Yeah it is necessary because the sign bit of `NaN` may or may not be positive, so `is_sign_positive` alone would skip negative `NaN`s and result in, say, `1NaNj` instead of `1+NaNj` as all `NaN`s are serialized as `NaN`. Python's `complex` behavior here:
```
>>> complex(1, float("nan"))
(1+nanj)
>>> complex(1, float("-nan"))
(1+nanj)
>>> math.copysign(1, float("nan"))
1.0
>>> math.copysign(1, float("-nan"))
-1.0
```
It's usually better to use [`signum`](https://docs.rs/num-traits/0.2.19/num_traits/float/trait.Float.html#tymethod.signum) which will propagate `NaN` instead, but since `+` is needed for `NaN`s as well we would have to check for `NaN` either way.

Comment (davidhewitt) [reply pydantic-core/src/serializers/type_serializers/complex.rs:86]: I see, so what you're saying is that while `1-NaNj` would be "most correct", looking at the Python complex output they chose to serialize negative NaN as positive NaN and so we should match that?

I guess some parsers may not accept it so it's better still to serialize it as `1+NaNj`?

That said, seems like `1-NaNj` would be fine for us:

```python
>>> math.copysign(0, pydantic.TypeAdapter(float).validate_python("-NAN"))
-0.0

>>> math.copysign(0, complex('1-NANj').imag)
-0.0
```

Comment (lhnwrk) [reply pydantic-core/src/serializers/type_serializers/complex.rs:86]: Ah, it's not exactly because Python chooses to serialize negative NaN as positive, but more so because [Rust opts to strip NaN sign during formatting](https://github.com/rust-lang/rust/blob/c7f5f3e0d5defe632d44743cbaed56272e2b67f0/library/core/src/num/flt2dec/mod.rs#L307). The same applies to Pydantic's float serializer:
```
>>> from pydantic import ConfigDict, TypeAdapter
>>> TypeAdapter(float, config=ConfigDict(ser_json_inf_nan="constants")).dump_json(float("-nan"))
b'NaN'
```
Hence, without the `is_nan()` check:
```
>>> TypeAdapter(complex).dump_json(complex("1+NaNj"))
b'"1+NaNj"'
>>> TypeAdapter(complex).dump_json(complex("1-NaNj"))
b'"1NaNj"'
```
And for Pydantic v2.12.5:
```
>>> TypeAdapter(complex).dump_json(complex("1+NaNj"))
b'"1NaNj"'
>>> TypeAdapter(complex).dump_json(complex("1-NaNj"))
b'"1NaNj"'
```

Since Pydantic's complex validator simply wraps around Python's complex constructor, negative NaN format is supported still:
```
>>> complex("1-NaNj")
(1+nanj)
>>> TypeAdapter(complex).validate_python("1-NaNj")
(1+nanj)
```

Comment (lhnwrk) [reply pydantic-core/src/serializers/type_serializers/complex.rs:86]: On "correctness", IEEE 754 specifies signed zeros, but intentionally leaves NaN signed bit and payload interpretation vague. So canonicalizing to "positive" NaN is still considered compliant. I'm okay with whatever behavior we want here, personally I was more concerned that some complex numbers cannot be round-trip serialized and deserialized because Python's complex constructor will raise a malformed string error on validation.

Comment (davidhewitt) [reply pydantic-core/src/serializers/type_serializers/complex.rs:86]: That makes total sense.

It looks like Python's `json` module serializes negative NaN without the sign, so I think we should continue to just drop the sign when serializing NaN. It seems there is no justification to changing the existing behaviour.

```python
>>> import json
>>> json.dumps(-float('nan'))
'NaN'
```

So it seems to me like what you have here is correct 👍

Comment (davidhewitt) [inline pydantic-core/src/serializers/type_serializers/complex.rs:86]: I think let's just document this, and as a _tiny_ nit, I believe that checking for `is_sign_positive` first is potentially more efficient (I guess the compiler might choose to reorder anyway).

```suggestion
        // NaN is always serialized as positive, e.g. matches Python json module
        if im.is_sign_positive() || im.is_nan() {
```

Comment (lhnwrk) [reply pydantic-core/src/serializers/type_serializers/complex.rs:86]: Added comments to clarify why the checks are needed! On the order or operations, unfortunately `is_sign_positive` does detect the actual sign bit of NaNs so they need to be handled as a special case first, otherwise "negative" NaN would skip the check even though they were serialized as "positive" 😬.

Comment (davidhewitt) [reply pydantic-core/src/serializers/type_serializers/complex.rs:86]: I think it's ok because `false || true` is still true, but no big stress. Let's merge. 👍
